crates/core/src/transport/received_packet_tracker.rs

```diff
@@ -0,0 +1,187 @@
+/// from/to a specific peer.
+///
+/// The caller must report when packets are received using the `report_received_packets` method.
+/// The caller must also call `get_receipts` periodically to check if any receipts need to be sent.
+///
+/// `get_receipts` should be called whenever a packet is sent.
+///
+/// `get_receipts` must **also** be called every `MAX_CONFIRMATION_DELAY` (50ms) and
+/// if the returned list is not empty, the list should be sent as receipts immediately in a noop.
+/// This may look something like this:
+///
+/// ```
+/// use std::thread::sleep;
+/// let mut received_packet_tracker = todo!();
+/// loop {
+///    let receipts = received_packet_tracker.get_receipts();
+///     if !receipts.is_empty() {
+///        // Send receipts in a noop message
+///     }
+///     sleep(MAX_CONFIRMATION_DELAY);
+///  }
+/// ```
+pub(super) struct ReceivedPacketTracker<T: TimeSource> {
+    pending_receipts: Vec<MessageId>,
+    message_id_time: VecDeque<(MessageId, Instant)>,
+    time_by_message_id: HashMap<MessageId, Instant>,
+    time_source: T,
+}
+
+impl ReceivedPacketTracker<SystemTime> {
+    pub(super) fn new() -> Self {
+        ReceivedPacketTracker {
+            pending_receipts: Vec::new(),
+            message_id_time: VecDeque::new(),
+            time_by_message_id: HashMap::new(),
+            time_source: SystemTime,
+        }
+    }
+}
+
+impl<T: TimeSource> ReceivedPacketTracker<T> {
+    pub(super) fn report_received_packets(&mut self, message_id: MessageId) -> ReportResult {
+        self.cleanup();
+        if self.time_by_message_id.contains_key(&message_id) {
+            ReportResult::AlreadyReceived
+        } else {
+            self.time_by_message_id
+                .insert(message_id, self.time_source.now());
+            self.message_id_time
+                .push_back((message_id, self.time_source.now()));
+
+            self.pending_receipts.push(message_id);
+
+            if self.pending_receipts.len() < MAX_PENDING_RECEIPTS {
+                ReportResult::Ok
+            } else {
+                ReportResult::QueueFull
+            }
+        }
+    }
+
+    /// Returns a list of packets that have been received since the last call to this function.
+    /// This should be called every time a packet is sent to ensure that receipts are sent
+    /// promptly. Every `MAX_CONFIRMATION_DELAY` (50ms) this should be called and if the returned
+    /// list is not empty, the list should be sent as receipts immediately in a noop packet.
+    pub(super) fn get_receipts(&mut self) -> Vec<MessageId> {
+        self.cleanup();
+
+        mem::take(self.pending_receipts.as_mut())
+    }
+
+    /// This function cleans up the `message_id_time` and `time_by_message_id` data structures.
+    /// It removes entries that are older than `RETAIN_TIME`.
+    fn cleanup(&mut self) {
+        let remove_before = self.time_source.now() - RETAIN_TIME;
+        while self
```

Comment:
if you use a btree then cleanup this can be done more efficiently , you can use instant as a key and get the range of ids which are older than RETAIN_TIME;

although then looking up by id won't be possible, but maybe you can still use a btree with id as key and retain by  the time value, since btrees are sorted it probably is faster.

heck i think a vec and using binary search may be faster for our use case, avoiding hashing altogether. worth experimenting a bit with this

Reply:
and probably you can avoid having both a hashmap and a vecdeque; since you are popping a range you can split a vec at the given index using binary search. I think this can be improved a bit